File: src/main/java/com/dumbdogdiner/Warrior/commands/arena/ArenaSpectateCommand.java
package com.dumbdogdiner.Warrior.commands.arena;

import com.dumbdogdiner.Warrior.Warrior;
import com.dumbdogdiner.Warrior.api.WarriorUser;
import com.dumbdogdiner.Warrior.api.arena.Arena;
import com.dumbdogdiner.Warrior.api.command.SubCommand;
import com.dumbdogdiner.Warrior.api.sesssions.ArenaSession;
import com.dumbdogdiner.Warrior.api.sesssions.GameState;
import com.dumbdogdiner.Warrior.managers.ArenaManager;
import com.dumbdogdiner.Warrior.managers.PlayerManager;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class ArenaSpectateCommand implements SubCommand {
    @Override
    public String getAlias() {
        return "spectate";
    }

    @Override
    public String getSyntax() {
        return "/arena spectate <arena>";
    }

    @Override
    public String getPermission() {
        return "warrior.arena.spectate";
    }

    @Override
    public boolean execute(CommandSender sender, String commandLabel, String[] args) {
        Arena a = ArenaManager.get(args[1]);
        WarriorUser u = PlayerManager.get(((Player)sender).getUniqueId());

        if(a == null || u == null) return false;

        new BukkitRunnable() {

            @Override
            public void run() {
                u.setSession(new ArenaSession(u.getUserId(), a, GameState.SPECTATING));
                u.setSpectating(true);
            }

        }.runTask(Warrior.getInstance());

        return true;
    }

    @Override
    public List<String> getArguments(CommandSender sender, String[] arguments) {
        List<String> strings = new ArrayList<>();
        if(arguments.length == 2)
            strings = ArenaManager.getArenas().stream()
                        .map(Arena::getName)
                        .collect(Collectors.toList());

        return strings;
    }
}

File: warrior/src/main/kotlin/com/dumbdogdiner/warrior/commands/arena/ArenaSpectateCommand.kt
package com.dumbdogdiner.warrior.commands.arena

import com.dumbdogdiner.warrior.api.command.SubCommand
import com.dumbdogdiner.warrior.api.arena.Arena
import com.dumbdogdiner.warrior.api.sessions.ArenaSession
import com.dumbdogdiner.warrior.api.sessions.GameState
import com.dumbdogdiner.warrior.Warrior
import com.dumbdogdiner.warrior.api.translation.Constants
import java.util.HashMap
import com.dumbdogdiner.warrior.api.util.TranslationUtil
import com.dumbdogdiner.warrior.user.User
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.bukkit.scheduler.BukkitRunnable
import java.util.ArrayList
import java.util.function.Function
import java.util.stream.Collectors

class ArenaSpectateCommand : SubCommand {
    override fun getAlias(): String {
        return "spectate"
    }

    override fun getSyntax(): String {
        return "/arena spectate <arena>"
    }

    override fun getPermission(): String {
        return "warrior.arena.spectate"
    }

    override fun execute(sender: CommandSender, commandLabel: String, args: Array<String>): Boolean {
        val a: Arena? = Warrior.arenaManager[args[1]]
        val u: User = Warrior.userCache[(sender as Player).uniqueId]
        if (a == null) return false

        if (u.session is ArenaSession) {
            if ((u.session as ArenaSession?)!!.state == GameState.IN_GAME) return false
        }

        object : BukkitRunnable() {
            override fun run() {
                val msg: String =
                    Warrior.translator.translate(Constants.Lang.ARENA_SPECTATE, object : HashMap<String, String>() {
                        init {
                            put("arena", args[1])
                        }
                    }, u)
                sender.sendMessage(TranslationUtil.getPrefix() + msg)
                u.setSession(ArenaSession(u.uniqueId, a, GameState.SPECTATING))
                u.isSpectating = true
            }
        }.runTask(Warrior.instance)
        return true
    }

    override fun getArguments(sender: CommandSender, arguments: Array<String>): List<String> {
        var strings: List<String> = ArrayList()
        if (arguments.size == 2) strings = Warrior.arenaManager.arenas
            .map<String>(Function { obj: Arena -> obj.name })
            .collect<List<String>, Any>(Collectors.toList<String>())
        return strings
    }
}